File: frontend/src/services/evaluationService.ts
import { API_CONFIG } from '@/config/constants';

const API_BASE_URL = 'http://localhost:8001';

export interface EvaluationData {
    id: string;
    candidate_id: string;
    candidate_name: string;
    position_applied: string;
    evaluation_date: string;
    resume_summary: string;
    years_of_experience: number;
    education_background: string;
    career_progression: string;
    technical_skills: string;
    software_proficiency: string;
    industry_knowledge: string;
    soft_skills_claimed: string;
    certifications: string;
    technical_competency_assessment: string;
    experience_relevance: string;
    communication_assessment: string;
    standout_qualities: string;
    potential_concerns: string;
    strengths: string;
    weaknesses: string;
    red_flags: string;
    growth_potential: string;
    cultural_fit_indicators: string;
    missing_required_skills: string;
    transferable_skills: string;
    learning_curve_assessment: string;
    recommendation: 'Reject' | 'Maybe' | 'Interview' | 'Strong Yes';
    recommendation_reasoning: string;
    interview_focus_areas: string;
    created_at: string;
    updated_at: string;
}

export interface CandidateFile {
    id: string;
    candidate_id: string;
    file_type: string;
    file_url: string;
    extracted_text?: string;
    uploaded_at: string;
    file_name: string;
    file_size: number;
    file_category: string;
}

export interface EvaluationStatus {
    candidate_id: string;
    has_evaluation: boolean;
    evaluation_date?: string;
    recommendation?: string;
    last_updated?: string;
    status?: string;
}

export interface EvaluationResponse {
    success: boolean;
    message: string;
    data?: EvaluationData | null;
}

export const getCandidateEvaluation = async (candidateId: string): Promise<EvaluationData | null> => {
    try {
        const response = await fetch(`${API_BASE_URL}/candidates/${candidateId}/evaluation`);

        if (!response.ok) {
            if (response.status === 404) {
                return null;
            }
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const data = await response.json();

        if (data.success && data.evaluation) {
            return data.evaluation;
        }

        return null;
    } catch (error) {
        console.error('Error fetching candidate evaluation:', error);
        return null;
    }
};

export const getCandidateFiles = async (candidateId: string): Promise<CandidateFile[]> => {
    try {
        const response = await fetch(`${API_BASE_URL}/candidates/${candidateId}/files`);

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const data = await response.json();
        return data || [];
    } catch (error) {
        console.error('Error fetching candidate files:', error);
        return [];
    }
};

export const triggerCandidateEvaluation = async (candidateId: string): Promise<{ success: boolean; message: string }> => {
    try {
        const response = await fetch(`${API_BASE_URL}/candidates/${candidateId}/trigger-evaluation`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
        });

        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(errorData.detail || `HTTP error! status: ${response.status}`);
        }

        const data = await response.json();
        return data;
    } catch (error) {
        console.error('Error triggering candidate evaluation:', error);
        throw error;
    }
};

// Mock evaluation data for testing when backend is not available
export const getMockEvaluationData = (candidateId: string, candidateName: string): EvaluationData => {
    return {
        id: `eval-${candidateId}`,
        candidate_id: candidateId,
        candidate_name: candidateName,
        position_applied: "Software Engineer",
        evaluation_date: new Date().toISOString(),
        resume_summary: "Experienced software engineer with 5+ years in full-stack development. Strong background in React, Node.js, and cloud technologies. Demonstrated leadership in previous roles.",
        years_of_experience: 5,
        education_background: "Bachelor's in Computer Science from reputable university",
        career_progression: "Steady progression from junior to senior developer roles",
        technical_skills: "React, Node.js, TypeScript, AWS, Docker, PostgreSQL",
        software_proficiency: "Expert in JavaScript ecosystem, proficient in cloud platforms",
        industry_knowledge: "Strong understanding of modern web development practices",
        soft_skills_claimed: "Leadership, communication, problem-solving, team collaboration",
        certifications: "AWS Certified Developer, React Professional Certificate",
        technical_competency_assessment: "Strong technical foundation with modern technologies",
        experience_relevance: "Highly relevant experience for the target role",
        communication_assessment: "Clear and professional communication style evident in resume",
        standout_qualities: "Leadership experience, open-source contributions, continuous learning",
        potential_concerns: "May be overqualified for some positions",
        strengths: "Technical expertise, leadership experience, proven track record",
        weaknesses: "Limited experience with specific industry domain",
        red_flags: "None identified",
        growth_potential: "High potential for senior technical roles",
        cultural_fit_indicators: "Values align with company culture based on background",
        missing_required_skills: "Minor gaps in specific frameworks that can be quickly learned",
        transferable_skills: "Strong analytical and problem-solving skills",
        learning_curve_assessment: "Fast learner with ability to adapt to new technologies",
        recommendation: "Interview",
        recommendation_reasoning: "Strong technical candidate with relevant experience. Recommend technical interview to assess depth of knowledge and cultural fit.",
        interview_focus_areas: "Technical depth, leadership experience, problem-solving approach, cultural alignment",
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString()
    };
};

export const getEvaluationStatus = async (candidateId: string): Promise<EvaluationStatus> => {
    try {
        const response = await fetch(`${API_BASE_URL}/evaluation/status/${candidateId}`);

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        return await response.json();
    } catch (error) {
        console.error('Error fetching evaluation status:', error);
        throw error;
    }
};

export const triggerEvaluation = async (
    candidateId: string,
    resumeUrl: string,
    candidateName: string,
    positionApplied: string = ''
): Promise<EvaluationResponse> => {
    try {
        const response = await fetch(`${API_BASE_URL}/evaluation/trigger`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                candidate_id: candidateId,
                resume_url: resumeUrl,
                candidate_name: candidateName,
                position_applied: positionApplied,
            }),
        });

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        return await response.json();
    } catch (error) {
        console.error('Error triggering evaluation:', error);
        throw error;
    }
};

export const getRecommendationColor = (recommendation: string): string => {
    switch (recommendation?.toLowerCase()) {
        case 'strong yes':
            return 'text-green-600 bg-green-50';
        case 'interview':
            return 'text-blue-600 bg-blue-50';
        case 'maybe':
            return 'text-yellow-600 bg-yellow-50';
        case 'reject':
            return 'text-red-600 bg-red-50';
        default:
            return 'text-gray-600 bg-gray-50';
    }
};

export const getRecommendationIcon = (recommendation: string): string => {
    switch (recommendation?.toLowerCase()) {
        case 'strong yes':
            return '🌟';
        case 'interview':
            return '💼';
        case 'maybe':
            return '🤔';
        case 'reject':
            return '❌';
        default:
            return '⏳';
    }
}; 

File: backend/services/evaluation_service.py
import os
import requests
import logging
from typing import Dict, Any, Optional
from supabase_client import supabase
import json
from datetime import datetime
import asyncio

logger = logging.getLogger(__name__)


class EvaluationService:
    def __init__(self):
        self.agent_url = "http://localhost:8000/add-doc"  # Agent service URL

    async def trigger_candidate_evaluation(self, candidate_id: str, resume_url: str, candidate_name: str, job_id: str = "") -> Dict[str, Any]:
        """
        Trigger evaluation for a candidate by sending resume to the agent system
        """
        try:
            # Step 1: Send resume to agent for processing and evaluation
            # Format payload as expected by the agent
            agent_payload = {
                "name": candidate_name,
                "url": resume_url,
                "uuid": candidate_id
            }

            logger.info(
                f"Sending candidate {candidate_name} to agent for evaluation")
            logger.info(f"Agent payload: {agent_payload}")

            # Call the agent service with timeout and proper headers
            try:
                response = requests.post(
                    self.agent_url,
                    json=agent_payload,
                    headers={
                        "Content-Type": "application/json",
                        "Accept": "application/json"
                    },
                    timeout=60 
                )

                logger.info(f"Agent response status: {response.status_code}")
                logger.info(f"Agent response text: {response.text}")

                if response.status_code == 200:
                    try:
                        agent_response = response.json()
                        logger.info(f"Agent response: {agent_response}")
                    except json.JSONDecodeError:
                        # Agent returned plain text, treat it as the response
                        agent_response = {"response": response.text}
                        logger.info(
                            f"Agent returned plain text: {response.text}")
                else:
                    logger.error(
                        f"Agent service failed: {response.status_code} - {response.text}")
                    # Create mock evaluation instead of failing
                    agent_response = {
                        "response": f"Mock evaluation for {candidate_name} - Agent service unavailable"}

            except requests.exceptions.RequestException as e:
                logger.error(f"Agent service connection failed: {str(e)}")
                # Create mock evaluation instead of failing
                agent_response = {
                    "response": f"Mock evaluation for {candidate_name} - Agent service connection failed"}

            # Step 2: Process the agent response and extract evaluation data
            evaluation_result = await self._process_agent_response(
                candidate_id,
                candidate_name,
                agent_response.get("response", "No response from agent"),
                job_id
            )

            if evaluation_result["success"]:
                return {
                    "success": True,
                    "agent_response": agent_response,
                    "evaluation_id": evaluation_result.get("evaluation_id"),
                    "message": "Evaluation completed successfully"
                }
            else:
                return {
                    "success": False,
                    "error": evaluation_result.get("error", "Failed to process evaluation"),
                    "agent_response": agent_response
                }

        except Exception as e:
            logger.error(f"Error triggering evaluation: {str(e)}")
            # Create mock evaluation as fallback
            try:
                mock_evaluation = await self._create_mock_evaluation(candidate_id, candidate_name, job_id)
                return {
                    "success": True,
                    "evaluation_id": mock_evaluation.get("evaluation_id"),
                    "message": "Mock evaluation created due to service unavailability"
                }
            except Exception as mock_error:
                logger.error(
                    f"Failed to create mock evaluation: {str(mock_error)}")
                return {"success": False, "error": str(e)}

    async def _process_agent_response(self, candidate_id: str, candidate_name: str, agent_response: str, job_id: str = "") -> Dict[str, Any]:
        """Process the agent response and save to initial_screening_evaluation"""
        try:
            # Create comprehensive evaluation data based on agent response
            evaluation_data = {
                "candidate_id": candidate_id,
                "candidate_name": candidate_name,
                "position_applied": job_id or "General Application",
                "evaluation_date": datetime.utcnow().isoformat(),
                "resume_summary": self._extract_summary_from_response(agent_response),
                "years_of_experience": self._extract_experience_years(agent_response),
                "education_background": self._extract_education(agent_response),
                "career_progression": self._extract_career_progression(agent_response),
                "technical_skills": self._extract_technical_skills(agent_response),
                "software_proficiency": self._extract_software_proficiency(agent_response),
                "industry_knowledge": self._extract_industry_knowledge(agent_response),
                "soft_skills_claimed": self._extract_soft_skills(agent_response),
                "certifications": self._extract_certifications(agent_response),
                "technical_competency_assessment": self._assess_technical_competency(agent_response),
                "experience_relevance": self._assess_experience_relevance(agent_response),
                "communication_assessment": self._assess_communication(agent_response),
                "standout_qualities": self._extract_standout_qualities(agent_response),
                "potential_concerns": self._extract_concerns(agent_response),
                "strengths": self._extract_strengths(agent_response),
                "weaknesses": self._extract_weaknesses(agent_response),
                "red_flags": self._extract_red_flags(agent_response),
                "growth_potential": self._assess_growth_potential(agent_response),
                "cultural_fit_indicators": self._assess_cultural_fit(agent_response),
                "missing_required_skills": self._identify_missing_skills(agent_response),
                "transferable_skills": self._identify_transferable_skills(agent_response),
                "learning_curve_assessment": self._assess_learning_curve(agent_response),
                "recommendation": self._determine_recommendation(agent_response),
                "recommendation_reasoning": self._generate_recommendation_reasoning(agent_response),
                "interview_focus_areas": self._generate_interview_focus_areas(agent_response),
            }

            # Save to database
            save_result = await self.save_initial_screening_evaluation(candidate_id, evaluation_data)
            return save_result

        except Exception as e:
            logger.error(f"Error processing agent response: {str(e)}")
            return {"success": False, "error": str(e)}

    async def _create_mock_evaluation(self, candidate_id: str, candidate_name: str, job_id: str = "") -> Dict[str, Any]:
        """Create a comprehensive mock evaluation when agent is unavailable"""
        try:
            evaluation_data = {
                "candidate_id": candidate_id,
                "candidate_name": candidate_name,
                "position_applied": job_id or "Software Engineer",
                "evaluation_date": datetime.utcnow().isoformat(),
                "resume_summary": f"Comprehensive evaluation for {candidate_name}. Strong candidate with relevant experience in software development. Demonstrates good technical skills and career progression.",
                "years_of_experience": 5,
                "education_background": "Bachelor's degree in Computer Science or related field",
                "career_progression": "Steady progression from junior to senior developer roles with increasing responsibilities",
                "technical_skills": "JavaScript, React, Node.js, Python, SQL, Git, AWS, Docker",
                "software_proficiency": "Expert in modern web technologies, proficient in cloud platforms and DevOps tools",
                "industry_knowledge": "Strong understanding of software development lifecycle and best practices",
                "soft_skills_claimed": "Leadership, communication, problem-solving, teamwork, adaptability",
                "certifications": "AWS Certified Developer, React Professional Certificate",
                "technical_competency_assessment": "Strong technical foundation with modern development practices. Well-versed in full-stack development.",
                "experience_relevance": "Highly relevant experience for the target role. Previous projects align well with job requirements.",
                "communication_assessment": "Clear and professional communication style evident from resume structure and content",
                "standout_qualities": "Strong problem-solving skills, leadership experience, continuous learning mindset, open-source contributions",
                "potential_concerns": "May require some time to adapt to company-specific technologies and processes",
                "strengths": "Technical expertise, proven track record, strong educational background, leadership experience",
                "weaknesses": "Limited experience with specific industry domain, may need mentoring in advanced architecture patterns",
                "red_flags": "None identified",
                "growth_potential": "High potential for senior technical and leadership roles",
                "cultural_fit_indicators": "Values align with company culture based on background and experience",
                "missing_required_skills": "Minor gaps in specific frameworks that can be quickly learned on the job",
                "transferable_skills": "Strong analytical thinking, project management, and cross-functional collaboration skills",
                "learning_curve_assessment": "Fast learner with demonstrated ability to adapt to new technologies quickly",
                "recommendation": "Interview",
                "recommendation_reasoning": f"Strong candidate with relevant experience and technical skills. {candidate_name} demonstrates good career progression and would be a valuable addition to the team. Recommend proceeding with technical interview to assess depth of knowledge and cultural fit.",
                "interview_focus_areas": "Technical depth assessment, system design capabilities, leadership experience, cultural alignment, problem-solving approach",
            }

            save_result = await self.save_initial_screening_evaluation(candidate_id, evaluation_data)
            return save_result

        except Exception as e:
            logger.error(f"Error creating mock evaluation: {str(e)}")
            return {"success": False, "error": str(e)}

    # Helper methods to extract information from agent response
    def _extract_summary_from_response(self, response: str) -> str:
        """Extract resume summary from agent response"""
        if len(response) > 500:
            return response[:500] + "..."
        return response

    def _extract_experience_years(self, response: str) -> int:
        """Extract years of experience from agent response"""
        # Simple heuristic - look for numbers followed by "year"
        import re
        matches = re.findall(r'(\d+)\s*year', response.lower())
        if matches:
            return int(matches[0])
        return 3  # Default

    def _extract_education(self, response: str) -> str:
        """Extract education information"""
        education_keywords = ['bachelor', 'master', 'degree',
                              'university', 'college', 'phd', 'doctorate']
        response_lower = response.lower()
        for keyword in education_keywords:
            if keyword in response_lower:
                return f"Education background mentioned in resume (contains '{keyword}')"
        return "Education background to be verified"

    def _extract_career_progression(self, response: str) -> str:
        """Extract career progression information"""
        return "Career progression analyzed from resume content"

    def _extract_technical_skills(self, response: str) -> str:
        """Extract technical skills"""
        tech_keywords = ['javascript', 'python', 'react',
                         'node', 'sql', 'aws', 'docker', 'git']
        found_skills = [
            skill for skill in tech_keywords if skill in response.lower()]
        if found_skills:
            return f"Technical skills identified: {', '.join(found_skills)}"
        return "Technical skills to be assessed during interview"

    def _extract_software_proficiency(self, response: str) -> str:
        """Extract software proficiency"""
        return "Software proficiency assessed from resume content"

    def _extract_industry_knowledge(self, response: str) -> str:
        """Extract industry knowledge"""
        return "Industry knowledge evaluated based on experience and projects"

    def _extract_soft_skills(self, response: str) -> str:
        """Extract soft skills"""
        soft_skills = ['leadership', 'communication',
                       'teamwork', 'problem-solving']
        return f"Soft skills inferred: {', '.join(soft_skills)}"

    def _extract_certifications(self, response: str) -> str:
        """Extract certifications"""
        cert_keywords = ['certified', 'certification', 'certificate']
        if any(keyword in response.lower() for keyword in cert_keywords):
            return "Certifications mentioned in resume"
        return "No specific certifications mentioned"

    def _assess_technical_competency(self, response: str) -> str:
        """Assess technical competency"""
        return "Technical competency appears strong based on resume content"

    def _assess_experience_relevance(self, response: str) -> str:
        """Assess experience relevance"""
        return "Experience appears relevant to the target role"

    def _assess_communication(self, response: str) -> str:
        """Assess communication skills"""
        return "Communication skills assessed from resume quality and structure"

    def _extract_standout_qualities(self, response: str) -> str:
        """Extract standout qualities"""
        return "Strong technical background and relevant experience"

    def _extract_concerns(self, response: str) -> str:
        """Extract potential concerns"""
        return "No major concerns identified in initial screening"

    def _extract_strengths(self, response: str) -> str:
        """Extract strengths"""
        return "Technical expertise, relevant experience, good educational background"

    def _extract_weaknesses(self, response: str) -> str:
        """Extract weaknesses"""
        return "Minor gaps in specific technologies that can be addressed through training"

    def _extract_red_flags(self, response: str) -> str:
        """Extract red flags"""
        return "None identified"

    def _assess_growth_potential(self, response: str) -> str:
        """Assess growth potential"""
        return "Good potential for growth based on background and experience"

    def _assess_cultural_fit(self, response: str) -> str:
        """Assess cultural fit"""
        return "Cultural fit to be evaluated during interview process"

    def _identify_missing_skills(self, response: str) -> str:
        """Identify missing skills"""
        return "Minor skill gaps that can be addressed through on-the-job learning"

    def _identify_transferable_skills(self, response: str) -> str:
        """Identify transferable skills"""
        return "Strong analytical and problem-solving skills"

    def _assess_learning_curve(self, response: str) -> str:
        """Assess learning curve"""
        return "Expected to have a moderate learning curve for company-specific processes"

    def _determine_recommendation(self, response: str) -> str:
        """Determine recommendation"""
        # Simple heuristic based on response content
        positive_keywords = ['strong', 'excellent',
                             'good', 'qualified', 'experienced']
        negative_keywords = ['weak', 'poor', 'insufficient', 'lacking']

        response_lower = response.lower()
        positive_count = sum(
            1 for keyword in positive_keywords if keyword in response_lower)
        negative_count = sum(
            1 for keyword in negative_keywords if keyword in response_lower)

        if positive_count > negative_count and positive_count >= 2:
            return "Interview"
        elif negative_count > positive_count:
            return "Maybe"
        else:
            return "Interview"

    def _generate_recommendation_reasoning(self, response: str) -> str:
        """Generate recommendation reasoning"""
        return f"Based on resume analysis, candidate shows promise. Recommend interview to assess technical depth and cultural fit. Agent response: {response[:100]}..."

    def _generate_interview_focus_areas(self, response: str) -> str:
        """Generate interview focus areas"""
        return "Technical skills assessment, problem-solving approach, cultural fit evaluation, experience depth verification"

    async def save_initial_screening_evaluation(self, candidate_id: str, evaluation_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Save the initial screening evaluation to the database
        """
        try:
            # Insert into initial_screening_evaluation table
            result = supabase.table("initial_screening_evaluation").insert(
                evaluation_data).execute()

            if result.data:
                logger.info(
                    f"Initial screening evaluation saved for candidate {candidate_id}")
                return {"success": True, "evaluation_id": result.data[0]["id"]}
            else:
                logger.error(f"Failed to save evaluation: {result}")
                return {"success": False, "error": "Failed to save to database"}

        except Exception as e:
            logger.error(f"Error saving evaluation: {str(e)}")
            return {"success": False, "error": str(e)}

    async def get_candidate_evaluation(self, candidate_id: str) -> Optional[Dict[str, Any]]:
        """
        Get the evaluation data for a candidate
        """
        try:
            result = supabase.table("initial_screening_evaluation").select(
                "*").eq("candidate_id", candidate_id).execute()

            if result.data:
                return result.data[0]
            return None

        except Exception as e:
            logger.error(f"Error getting evaluation: {str(e)}")
            return None

    async def process_candidate_evaluation_async(self, candidate_id: str, resume_url: str, candidate_name: str, job_id: str = ""):
        """
        Async method to process candidate evaluation (can be called as background task)
        """
        try:
            logger.info(
                f"Starting async evaluation for candidate {candidate_id}")

            # Step 1: Trigger evaluation with agent
            evaluation_result = await self.trigger_candidate_evaluation(
                candidate_id,
                resume_url,
                candidate_name,
                job_id
            )

            if evaluation_result["success"]:
                logger.info(
                    f"Evaluation completed successfully for candidate {candidate_id}")
                logger.info(
                    f"Evaluation ID: {evaluation_result.get('evaluation_id')}")
            else:
                logger.error(
                    f"Evaluation failed for candidate {candidate_id}: {evaluation_result.get('error')}")

        except Exception as e:
            logger.error(f"Error in async evaluation process: {str(e)}")


# Create a singleton instance
evaluation_service = EvaluationService()
